# 0. Import and Install Libraries

## Installs

In [2]:
# !pip install yellowbrick

In [ ]:
# !pip install graphviz

## Imports

In [8]:
import pandas as pd

In [9]:
import numpy as np

# 1. Research Goals

## Problem statement
This project investigates which factors most strongly influence housing prices in Melbourne and whether housing prices can be predicted reliably using property and location data.

## Context
The Melbourne housing market is influenced by both structural property features and geographic location. Understanding which factors drive housing prices is important for buyers, sellers, and analysts. This dataset makes it possible to examine price patterns and test whether housing prices can be predicted from available variables.

## Goals
The goal of this project is to investigate the main determinants of housing prices in Melbourne and to build a predictive model that estimates house prices based on housing characteristics and location.

## Research questions
* Which factors are the strongest predictors of housing price?
* How does location affect housing prices across Melbourne?
* How do housing prices differ between property types?
* To what extent can machine learning predict housing prices accurately?

## Hypotheses
* Housing prices increase with size, number of rooms, and number of bathrooms.
* Properties located closer to the city center are more expensive.
* Houses are more expensive than units and townhouses.
* Location-based features contribute significantly to price prediction.
* A machine learning model can predict housing prices with useful accuracy.


# 2. Retrieve Data

## 2.1 Melbourne Data

The data is originally scraped from Domain.com.au. and the uploaded to Kaggle.com

In [12]:
df_melb = pd.read_csv(r"C:\Users\Administrator\BI\Melbourne\data\melb_data.csv")

## Data Dictionary
| Variable | Description |
| :--- | :--- |
| **Rooms** | Number of rooms |
| **Price** | Price in dollars |
| **Method** | S - property sold; SP - property sold prior; PI - property passed in; PN - sold prior not disclosed; SN - sold not disclosed; NB - no bid; VB - vendor bid; W - withdrawn prior to auction; SA - sold after auction; SS - sold after auction price not disclosed. N/A - price or highest bid not available. |
| **Type** | br - bedroom(s); h - house, cottage, villa, semi, terrace; u - unit, duplex; t - townhouse; dev site - development site; o res - other residential. |
| **SellerG** | Real Estate Agent |
| **Date** | Date sold |
| **Distance** | Distance from CBD |
| **Regionname** | General Region (West, North West, North, North east …etc) |
| **Propertycount** | Number of properties that exist in the suburb. |
| **Bedroom2** | Scraped # of Bedrooms (from different source) |
| **Bathroom** | Number of Bathrooms |
| **Car** | Number of carspots |
| **Landsize** | Land Size |
| **BuildingArea** | Building Size |
| **CouncilArea** | Governing council for the area |

## 2.2 Denmark Data

# 3. Data Preparation

## Melbourne Data

### 3.1 Data Inspection

Quick inspection of the dataset, to see what needs cleaning

In [19]:
# Check rows and columns
df_melb.shape

(13580, 21)

In [23]:
# Random samples to get a quick overview of the data
df_melb.sample(5)

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,...,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
925,Bentleigh East,2/16 Bessie St,3,u,770000.0,PI,hockingstuart,25/02/2017,13.9,3165.0,...,2.0,1.0,281.0,132.0,2005.0,Glen Eira,-37.92570,145.08110,Southern Metropolitan,10969.0
164,Altona,164 Maidstone St,3,h,635000.0,S,Sweeney,14/05/2016,13.8,3018.0,...,1.0,4.0,625.0,NaN,1960.0,Hobsons Bay,-37.86110,144.81140,Western Metropolitan,5301.0
3845,Malvern East,349 Waverley Rd,5,h,1810000.0,PI,RT,10/09/2016,11.2,3145.0,...,3.0,2.0,751.0,NaN,1934.0,Stonnington,-37.87750,145.06220,Southern Metropolitan,8801.0
12396,Roxburgh Park,23 Wheatley Av,3,h,515000.0,S,Barry,3/09/2017,20.6,3064.0,...,2.0,2.0,532.0,NaN,2000.0,NaN,-37.62554,144.92006,Northern Metropolitan,5833.0
10066,Templestowe Lower,23 Gertrude St,4,h,1352000.0,S,Ray,24/06/2017,12.4,3107.0,...,2.0,3.0,725.0,NaN,NaN,Manningham,-37.76472,145.10261,Eastern Metropolitan,5420.0


In [21]:
# Check the datatypes
df_melb.dtypes

Suburb               str
Address              str
Rooms              int64
Type                 str
Price            float64
Method               str
SellerG              str
Date                 str
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea          str
Lattitude        float64
Longtitude       float64
Regionname           str
Propertycount    float64
dtype: object

In [27]:
# Check to see how many unique values there is for each column, to see if some columns can be dropped
df_melb.nunique()

Suburb             314
Address          13378
Rooms                9
Type                 3
Price             2204
Method               5
SellerG            268
Date                58
Distance           202
Postcode           198
Bedroom2            12
Bathroom             9
Car                 11
Landsize          1448
BuildingArea       602
YearBuilt          144
CouncilArea         33
Lattitude         6503
Longtitude        7063
Regionname           8
Propertycount      311
dtype: int64

In [37]:
# Check to see if there are duplicated rows
df_melb.duplicated().sum()

np.int64(0)

In [34]:
# Check to see if there are missing values
df_melb.isna().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                62
Landsize            0
BuildingArea     6450
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

In [30]:
# Check of the descriptive statistics to check for odd values
df_melb.describe()

,Rooms,Price,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,Lattitude,Longtitude,Propertycount
count,13580.000000,1.358000e+04,13580.000000,13580.000000,13580.000000,13580.000000,13518.000000,13580.000000,7130.000000,8205.000000,13580.000000,13580.000000,13580.000000
mean,2.937997,1.075684e+06,10.137776,3105.301915,2.914728,1.534242,1.610075,558.416127,151.967650,1964.684217,-37.809203,144.995216,7454.417378
std,0.955748,6.393107e+05,5.868725,90.676964,0.965921,0.691712,0.962634,3990.669241,541.014538,37.273762,0.079260,0.103916,4378.581772
min,1.000000,8.500000e+04,0.000000,3000.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1196.000000,-38.182550,144.431810,249.000000
25%,2.000000,6.500000e+05,6.100000,3044.000000,2.000000,1.000000,1.000000,177.000000,93.000000,1940.000000,-37.856822,144.929600,4380.000000
50%,3.000000,9.030000e+05,9.200000,3084.000000,3.000000,1.000000,2.000000,440.000000,126.000000,1970.000000,-37.802355,145.000100,6555.000000
75%,3.000000,1.330000e+06,13.000000,3148.000000,3.000000,2.000000,2.000000,651.000000,174.000000,1999.000000,-37.756400,145.058305,10331.000000
max,10.000000,9.000000e+06,48.100000,3977.000000,20.000000,8.000000,10.000000,433014.000000,44515.000000,2018.000000,-37.408530,145.526350,21650.000000


#### Data Inspection Findings

We have observed the following that needs further examination and/or cleaning:

* 'Rooms' (from original source) vs 'Bedroom2' (from other source)

   Both colums contains nearly the same means (2.94 vs 2.91) and identical quartiles. However we can also see that in 'Bedroom2' the min value is 0       rooms and the max is 20 rooms vs 1 and 10 in 'Rooms', so there are some clear outliers in 'Bedroom2'. So to avoid redundant data we will drop the      'Bedroom2' column and keep the data that was scraped the same place as all the other columns.
  
* Possible Outliers and odd values
  -  Some properties have 0 'Bathrooms', which is possible with fx. shared bathrooms, but it seems unlikely.
  -  Some properties have a 'Landsize' of 0 while the max value is 433 014 m2.
  -  Some properties have a 'BuildingArea' of 0 while the max value is 44 515 m2
  -  There is a property built in year 1196, which is highly unlikely in Austrailia as europeans arrived in 1606 and therefore likely an error


* Columns with missing values:
  - 'Car' has 62 missing values
  - 'BuildingArea' has 6450 missing values
  - 'YearBuilt' has 5375 missing values
  - 'CouncilArea' has 1269 missing values

### 3.2 Data Cleaning

In [75]:
df_melb_clean = df_melb.copy()

#### Drop column 'Bedroom2'


In [76]:
df_melb_clean.drop(columns='Bedroom2')

,Suburb,Address,Rooms,Type,Price,Method,SellerG,Date,Distance,Postcode,Bathroom,Car,Landsize,BuildingArea,YearBuilt,CouncilArea,Lattitude,Longtitude,Regionname,Propertycount
0,Abbotsford,85 Turner St,2,h,1480000.0,S,Biggin,3/12/2016,2.5,3067.0,1.0,1.0,202.0,NaN,NaN,Yarra,-37.79960,144.99840,Northern Metropolitan,4019.0
1,Abbotsford,25 Bloomburg St,2,h,1035000.0,S,Biggin,4/02/2016,2.5,3067.0,1.0,0.0,156.0,79.0,1900.0,Yarra,-37.80790,144.99340,Northern Metropolitan,4019.0
2,Abbotsford,5 Charles St,3,h,1465000.0,SP,Biggin,4/03/2017,2.5,3067.0,2.0,0.0,134.0,150.0,1900.0,Yarra,-37.80930,144.99440,Northern Metropolitan,4019.0
3,Abbotsford,40 Federation La,3,h,850000.0,PI,Biggin,4/03/2017,2.5,3067.0,2.0,1.0,94.0,NaN,NaN,Yarra,-37.79690,144.99690,Northern Metropolitan,4019.0
4,Abbotsford,55a Park St,4,h,1600000.0,VB,Nelson,4/06/2016,2.5,3067.0,1.0,2.0,120.0,142.0,2014.0,Yarra,-37.80720,144.99410,Northern Metropolitan,4019.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
13575,Wheelers Hill,12 Strada Cr,4,h,1245000.0,S,Barry,26/08/2017,16.7,3150.0,2.0,2.0,652.0,NaN,1981.0,NaN,-37.90562,145.16761,South-Eastern Metropolitan,7392.0
13576,Williamstown,77 Merrett Dr,3,h,1031000.0,SP,Williams,26/08/2017,6.8,3016.0,2.0,2.0,333.0,133.0,1995.0,NaN,-37.85927,144.87904,Western Metropolitan,6380.0
13577,Williamstown,83 Power St,3,h,1170000.0,S,Raine,26/08/2017,6.8,3016.0,2.0,4.0,436.0,NaN,1997.0,NaN,-37.85274,144.88738,Western Metropolitan,6380.0
13578,Williamstown,96 Verdon St,4,h,2500000.0,PI,Sweeney,26/08/2017,6.8,3016.0,1.0,5.0,866.0,157.0,1920.0,NaN,-37.85908,144.89299,Western Metropolitan,6380.0


#### 3.3 Missing values

##### 'Cars'
For the 62 missing car values we will use the median to fill the gaps as you can not have 1.4 car spaces

In [77]:
median_cars = df_melb_clean['Car'].median()
df_melb_clean['Car'] = df_melb_clean['Car'].fillna(median_cars)

##### 'BuildingArea'

In [78]:
# First we will check that the rows with missing BuildingArea are not empty lots without buildings (development sites)
df_melb_clean['Type'].unique()

<ArrowStringArray>
['h', 'u', 't']
Length: 3, dtype: str

In [79]:
# Check to see how the NaN's are distributed among the types of properties
df_melb_clean[df_melb_clean['BuildingArea'].isna()]['Type'].value_counts()

Type
h    4589
u    1433
t     428
Name: count, dtype: int64

So we can conclude that the rows with missing values are not just empty lots without houses.
Given that nearly half of the rows are missing the BuildingArea we will not drop it but fill it with the median for similar properties based on their type and number of rooms:

In [83]:
# Calculate the median for each group
grouped_median = df_melb_clean.groupby(['Type', 'Rooms'])['BuildingArea'].transform('median')

# Fill in the missing values, first with the grouped median and then with a global median if there are still missing values
df_melb_clean['BuildingArea'] = df_melb_clean['BuildingArea'].fillna(grouped_median).fillna(df_melb_clean['BuildingArea'].median())

##### YearBuilt

In [ ]:
# TODO Mean for yearbuilt to replace 

##### CouncilArea

In [ ]:
# TODO Map Regionname til council hvis vi kan google os frem til hvad det skal være

In [84]:
# Check to see if there are still missing values
df_melb_clean.isna().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
YearBuilt        5375
CouncilArea      1369
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

#### 3.4 Removing Outliers

#### 3.5 Feature Engineering

In [87]:
# Calculate price per m2 for each row
df_melb_clean["Price_per_m2"] = df_melb_clean["Price"] / df_melb_clean["BuildingArea"]

In [88]:
# Calculate the age for each row
df_melb_clean["Age"] = 2026 - df_melb_clean["YearBuilt"]

#### 3.6 Data Transformation

# 4. Data Exploration

## 4.1 EDA

## 4.2 Correlations

## 4.3 Descriptive Statistics

## 4.4 Inferential Statistics

# 5. Data Modelling

# 6. Presentation